File: Bowling/src/classes/BaseFrame.cs
namespace Bowling;

public abstract class BaseFrame
{
  public int BonusPoints { get; set; }
  public Dictionary<string, List<Pin>> KnockedDownPins { get; set; }
  public int Number { get; set; }
  public List<Pin> StandingPins { get; set; }
  public int TotalScore { get; set; }

  public BaseFrame(int Number)
  {
    BonusPoints = 0;
    KnockedDownPins = new()
    {
        { "FirstThrow", new() },
        { "SecondThrow", new() },
    };
    if (Number == 10)
    {
      KnockedDownPins.Add("ThirdThrow", new());
    }
    this.Number = Number;
    StandingPins = new();
    TotalScore = 0;
  }

  public int FirstThrowScore
  {
    get => KnockedDownPins["FirstThrow"].Count;
  }
  public int FrameScore
  {
    get => RoundScore + BonusPoints;
  }
  public virtual int RoundScore
  {
    get => KnockedDownPins["FirstThrow"].Count + KnockedDownPins["SecondThrow"].Count;
  }
  public int SecondThrowScore
  {
    get => KnockedDownPins["SecondThrow"].Count;
  }
}

File: src/classes/base_frame.py
from .pin import Pin

class BaseFrame:
  def __init__(self, number: int):
    self.bonus_points: int = 0
    self.knock_down_pins: dict[str, list[Pin]] = {
      'first_throw': [],
      'second_throw': [],
    }
    if number == 10:
      self.knock_down_pins['third_throw'] = []
    self.number: int = number
    self.standing_pins: list[Pin] = []
    self.total_score: int = 0

  @property
  def first_throw_score(self) -> int:
    return len(self.knock_down_pins['first_throw'])

  @property
  def frame_score(self) -> int:
    return self.round_score + self.bonus_points

  @property
  def round_score(self) -> int:
    return len(self.knock_down_pins['first_throw']) + len(self.knock_down_pins['second_throw'])

  @property
  def second_throw_score(self) -> int:
    return len(self.knock_down_pins['second_throw'])
